chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+
+
+class Recipe:
+    """
+    A recipe is the workflow representation of a simulation. It is struktured
+    as a list of nodes of class Node. The nodes already contain information
+    about the data flow and dependencies in the workflow, so there is not
+    explicit representation of edges or dependencies needed.
+    """
+    nodes = []
+
+    def __init__(self, nodelist):
+        """
+        Initialises a recipe by appending the `nodelist` to `nodes`.
+
+        :param nodelist: list of simulation steps (nodes)
+        :type nodelist: Node[]
+        """
+        self.nodes = nodelist
+        # Making sure, that nodelist is a list of type Node
+        try:
+            emptyNode = Node(
+                "nodeName",
+                {"i": "input"},
+                {"o": "output"},
+                "collect")
+            self.nodes.append(emptyNode)
+            self.nodes.pop()
+        except Exception as exc:         # mit Klasse
+            logger.exception(exc)
+            #raise TypeError("""Class Recipe expects a list to be
+            #    initialised with.""")
+
+    def inputIntegrity(self):
+        """
+        Tests if there is exactly one incoming edge to every input.
+        Warns, if there is no incoming edge. Excludes nodes from recipe,
+        that have no incoming edge for a node or have uncomputable inputs
+        because of missing inputs in parent nodes.
+        
+        :returns:
+            err - String. None if correct. Else error message. \n
+            warn - String. None if correct. Else error message.
+        """
+        err = None
+        warn = None
+        # 1. make unique list of outputs
+        outputs_of_all_nodes = []
+        for node in self.nodes:
+            for key in node.outputs:
+                output = node.outputs[key]
+                if output in outputs_of_all_nodes:
+                    # overwrite exception class to generate warns
+                    err = (("" if err is None else err) + 'The output ' +
+                           output + ' of node ' + node.name +
+                           ' has the same name as an output declared before. ')
+                else:
+                    outputs_of_all_nodes.append(output)
+        # 2. see if inputs are from flavour, are file paths to existing files
+        # or are in output list
+        try_again = True
+        while try_again:
+            unreachable_nodes = []
+            for node in self.nodes:
+                nodeIsValid = True
+                for key in node.inputs:
+                    input = node.inputs[key]
+                    # to do: Werte direkt zulassen, nicht nur ?ber flavour
+                    # WARN schmei?en
+                    # python logs verwenden: kein flavour, kein output, also
+                    # interpretiert als string
+                    # chef analyse (from log)
+                    inputIsValid = False
+                    prefix = os.path.splitext(input)[0]
```

Comment:
I suspect this kind of string parsing will be done more often. If there are more than 1 places, encapsulating it into functions might be preferrable.

Further, since that is how the namespaces unfold, I strongly could imagine that defining a class, that handles namespaces transparently, could be very helpful. Let alone by the possibility to define checkups, conditioning or other side-tasks as methods directly into that structure.

Maybe collect requirements for the use of namespaces first and let's discuss them :)